Add Features in New Environment & Move Notebooks

louisowen6 opened the pull request:
Add Features in `env_stocktrading_cashpenalty.py`:
1) Split `transaction_cost_pct` to `buy_cost_pct` and `sell_cost_pct`
2) Add optional `actions` discretization
3) Incorporate different length of data in multistock trading
4) Minor fixing in class docs
5) Activate `turbulence` 
6) Add `currency` parameter
7) Enrich logging information

Others:
- Adjust `env_multistock_cashpenalty.ipynb`
- Move all notebooks to 'notebooks' folder

spencerR1992 commented on finrl/env/env_stocktrading_cashpenalty.py:
```diff
@@ -247,27 +261,36 @@ def step(self, actions):
             # multiply action values by our scalar multiplier and save
-            actions = actions * self.hmax
+            actions = actions * self.hmax 
+            #buy/sell only if the price is > 0 (no missing data in this particular date 
```
is it possible that prices could be lt 0? maybe a dumb question?

louisowen6 replied:
No it can't be, but in the previous work on incorporating different lengths of data in multi stock trading env, we agreed to do the imputation of missing date with 0. Please see this [notebook ](https://github.com/AI4Finance-LLC/FinRL-Library/blob/master/FinRL_ensemble_multiple_stock_trading_ICAIF_2020.ipynb) to know more

spencerR1992 replied:
ohhhh, I see. 
Ok, well, in that case...
@BruceYanghy @XiaoYangLiu-FinRL we might need to consider making a config variable for missing value (like np.nan or None) so we don't mix up 'no data' for '0 data'.

XiaoYangLiu-FinRL replied:
Yes, we previously discussed about adding zeros at the beginning to deal with missing values (not on the market yet).

spencerR1992 commented on finrl/env/env_stocktrading_cashpenalty.py:
```diff
@@ -247,27 +261,36 @@ def step(self, actions):
-            actions = actions * self.hmax
+            actions = actions * self.hmax 
+            #buy/sell only if the price is > 0 (no missing data in this particular date 
+            actions = np.array([actions[i] if closings[i]>0 else 0 for i in range(len(actions))])
```
there's a numpy way to do this, let's do that instead of the list comp. 
I believe it might be np.where

louisowen6 replied:
Sure

spencerR1992 commented on finrl/env/env_stocktrading_cashpenalty.py:
```diff
@@ -247,27 +261,36 @@ def step(self, actions):
 
-            # scale cash purchases to asset # changes
-            actions = actions / closings
+            # scale cash purchases to asset
+            actions = actions // closings
+            #convert into integer because we can't buy fraction of shares
+            actions = (actions.astype(int))
```
@louisowen6 I'd like this discretization of actions to be optional (i.e. )
`if self.discrete:-> do discretization of actions'`

spencerR1992 replied:
full disclosure, I plan to trade with this tool, but trade on an exchange that has fractional shares  (like robinhood)

louisowen6 replied:
Sure! Never thought we can trade in fractional shares. This is new knowledge for me :)

spencerR1992 reviewed (changes requested):
Louis, 
Nice work! My biggest gripe here is about discretization. I'd ask that you make that an optional step, as I mentioned. 
Also, that one listcomp needs to be converted to a numpy op.

louisowen6 commented:
Other updates:

1) Added `min_shares` parameter to control the minimum number of shares that can be bought in each trade (only for scaling)
2) Added `negative_asset_penalty` to further control the flexibility of the reward function
3) Enrich logging information
4) Fix saving `actions_memory` value. Previous implementation output bigger actions value than the number of assets we have

spencerR1992 commented on finrl/env/env_stocktrading_cashpenalty.py:
```diff
@@ -208,7 +236,7 @@ def get_reward(self):
             cash = self.account_information['cash'][-1]
             cash_penalty = max(0, (assets*self.cash_penalty_proportion-cash))
-            assets -= cash_penalty
+            assets -= cash_penalty if assets > self.initial_amount else (self.negative_asset_penalty+1)*cash_penalty
```
Sorry, not sure I'm following the logic here;
Why do we need extra penalty we are negative? This seems like we're inserting a human logic fallacy into the mix.

spencerR1992 replied:
Also, maybe I'm wrong here, so in that case, it might be good for us to have an easy way to choose a reward function or toggle between reward functions?

louisowen6 replied:
The idea is to give a bigger penalty when we are short of cash and our current total asset is less than the initial amount. Yes, actually we can just turn that penalty off by assigning the `negative_asset_penalty` with 0

louisowen6 replied:
Hey, how about we create a parameter for choosing the type of risk a user's can take (e.g. risk-averse, risk-moderate, risk-taker). What do you think? So it will be based on this `negative_asset_penalty` parameter.

cc: @XiaoYangLiu-FinRL @BruceYanghy

spencerR1992 replied:
I think this is an anthropomorphic assumption that I would want to see proven out. 
From a purely economics frame, the weighting of negative and positive consequences should allow the model to balance risk vs. reward properly. Intuitively it might be the case that adding steeper penalty for negative returns yields positive returns more often. However, there are a lot of variables at play, so we'd need to prove that. The common fallacy name is that 'losses loom larger than gains' which is why humans often get out of positions that they should hold when losses come in, and why people who don't invest their own money are often the most effective stewards of wealth. 
https://www.behavioraleconomics.com/resources/mini-encyclopedia-of-be/loss-aversion/#:~:text=Loss%20aversion%20is%20an%20important,as%20the%20pleasure%20of%20gaining.

Along these lines, I had actually contemplated adding a reward term related to sharpe ratio at any given step, or to concentration of holdings. However, my backtesting results have been very encouraging (>20% annual returns, >0.9 sharpe avg, beat market by 5% annual) so I'm inclined to leave this alone. 

For now, I do not agree with changing the reward function in any way given the complexity, and the unknowns. 
I'd suggest monkey patching the environment prior to training to make the change you're desiring here.

louisowen6 replied:
Okay! I'll remove this in the next commit

spencerR1992 replied:
Thank you! Let's plan to collaborate on ideas of how to make the reward function configurable in the future!

spencerR1992 commented on finrl/env/env_stocktrading_cashpenalty.py:
```diff
@@ -247,27 +275,40 @@ def step(self, actions):
             actions = np.maximum(actions, -np.array(holdings))
+
             self.transaction_memory.append(actions)
+            self.actions_memory.append(actions*closings)
```
Can we move this up so it's not subject to any of our modifications? Actions_memory is supposed to capture what the model is trying to do, where transaction memory indicates what it could do.

louisowen6 replied:
Ahh I see, okay sure

spencerR1992 commented on finrl/env/env_stocktrading_cashpenalty.py:
```diff
@@ -175,28 +198,33 @@ def log_step(self, reason, terminal_reward=None):
             self.account_information["cash"][-1]
             / self.account_information["total_assets"][-1]
         )
+        GL_pct = self.account_information["total_assets"][-1] / self.initial_amount
```
lowercase variables here.

louisowen6 replied:
Sure

spencerR1992 commented on finrl/env/env_stocktrading_cashpenalty.py:
```diff
@@ -52,28 +56,38 @@ class StockTradingEnvCashpenalty(gym.Env):
+        self.df = self.df.set_index(date_col_name
+        self.min_shares = min_shares
+        self.hmax = hmax / min_shares
+        self.initial_amount = initial_amount / min_shares
```
what is the purpose of min_shares?

louisowen6 replied:
In some countries, including Indonesia, we can only buy with a minimum of 100 shares. So this acts as a scaling factor

spencerR1992 replied:
Ok, in that case, I don't think that scaling initial_amt by min_shares is correct.

louisowen6 replied:
Hmm, could you please elaborate more on how should we tackle this problem in a simpler solution rather than scaling by min_shares?

spencerR1992 replied:
ok. Initial amt is the amount of cash you start with. even if you can trade only in increments of 100 shares, it should not change the initial amount of cash but only the numbers of shares traded at each step.

louisowen6 replied:
Yes, I agree with you. However, to simplify the work, I think we can just scale down the `initial_amount` and `hmax` by `min_shares`, and think that during each trade, when we buy 1 share is equal to what we buy `min_shares` of shares in real life

spencerR1992 replied:
I don't agree with this. 
The cash should not be impacted by the min shares. It's ok if this is a requirement for the actions space, but a user putting in a initial_cash amount would be surprised to see that their cash on hand had been cut by the factor `min_shares` at the first step and be very confused.

XiaoYangLiu-FinRL replied:
ooh, the account balance is a purse, managing only "cash in" and "cash out", right?  
We may consider to accommodate several cases: 1. stock shares in integer;  2. stock shares in unit of "100 shares";  3. btc allows fractional shares;  4. Robinhood may allow cash and make it into fractional shares.  Or other cases that I am not aware of.

louisowen6 replied:
@XiaoYangLiu-FinRL Yes, I agree we have to accommodate all of those possibilities.

@spencerR1992 That's a valid point. That's why in the logging I multiply it back with the `min_cash` parameter. 

```
logger.record("environment/total_assets", int(self.min_shares*self.account_information['total_assets'][-1]))
```

CMIIW, I think if we only scaled-down the `hmax` without also scaling down the `initial_amount` then it will be not fair...

louisowen6 commented:
Removed `negative_cash_penalty` parameter
